dataprofiler/tests/profilers/test_categorical_column_profile.py

```diff
@@ -936,6 +1140,93 @@ def test_json_decode_after_update(self):
 
         test_utils.assert_profiles_equal(deserialized, expected_profile)
 
+    def test_cms_heavy_hitters_threshold(self):
+
+        df_categorical = pd.Series(
+            [
+                "a",
+                "a",
+                "a",
+                "a",
+                "a",
+                "b",
+                "b",
+                "b",
+                "b",
+                "b",
+                "c",
+                "c",
+                "c",
+                "c",
+                "c",
+                "c",
+                "c",
+                "c",
+                "c",
+                "c",
+            ]
+        )
+        options = CategoricalOptions()
+        options.cms = True
+        options.cms_confidence = 0.95
+        options.cms_relative_error = 0.01
+        options.heavy_hitters_threshold = 2
+
+        profile = CategoricalColumn("test_name", options)
+        profile.update(df_categorical)
+
+        self.assertEqual(1, len(profile.categories))
+        self.assertTrue(profile.sample_size >= 10)
+
+    def test_cms_update_hybrid_batch_stream(self):
+        dataset = pd.Series(["a"] * 4 + ["b"] * 12 + ["c"] * 14)
+        dataset1 = pd.Series(["a"] * 9 + ["b"] * 11 + ["c"] * 9 + ["d"] * 1)
+
+        options = CategoricalOptions()
+        options.cms = True
+        options.cms_confidence = 0.95
+        options.cms_relative_error = 0.01
+        options.heavy_hitters_threshold = 3
+
+        profile = CategoricalColumn("test_name", options)
+        profile.update(dataset)
+        profile.update(dataset1)
+        expected_categories = ["b", "c"]
+
+        self.assertEqual(profile.sample_size, len(dataset) + len(dataset1))
+
+        self.assertTrue(profile._categories["c"] >= 23)
+
+        self.assertCountEqual(expected_categories, profile.categories)
+
+        self.assertTrue(profile._categories["b"] >= 23)
+
+    def test_cms_profile_merge_via_add(self):
+
+        dataset = pd.Series(["a"] * 9 + ["b"] * 12 + ["c"] * 9)
+        dataset1 = pd.Series(["a"] * 6 + ["b"] * 10 + ["c"] * 14)
+
+        expected_categories = ["b", "c"]
+        options = CategoricalOptions()
+        options.cms = True
+        options.cms_confidence = 0.95
+        options.cms_relative_error = 0.01
+        options.heavy_hitters_threshold = 3
+
+        profile1 = CategoricalColumn("test_name", options)
+        profile1.update(dataset)
+        profile2 = CategoricalColumn("test_name", options)
+        profile2.update(dataset1)
+
+        # Add profiles
+        profile3 = profile1 + profile2
+        self.assertCountEqual(expected_categories, profile3.categories)
+        self.assertEqual(
+            profile3.sample_size, profile1.sample_size + profile2.sample_size
+        )
+        self.assertTrue(profile3._categories["b"] == 22)
+        self.assertTrue(profile3._categories["c"] == 14)
```

Comment:
there is indeed a bug here - but our HH pruning actually removes the C category completely right now.

Reply:
the intent of this test was to try and illustrate the "gap" that occurs during add. Since we aren't treating as a stream, missing_heavy_hitters_dict doesn't help us capture the "c":9 from the first dataset that fell below the "batch HH" threshold...

I originally thought it would be 14 but realize now that by the same logic, 14 < (total_samples/HH)